src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs

```diff
@@ -594,39 +419,35 @@ public override string GetDocumentationCommentXml(CultureInfo preferredCulture =
 
         #endregion
 
-        public override ImmutableArray<CustomModifier> ReturnTypeCustomModifiers
-        {
-            get
-            {
-                return ImmutableArray<CustomModifier>.Empty;
-            }
-        }
+        public override ImmutableArray<CustomModifier> ReturnTypeCustomModifiers => ImmutableArray<CustomModifier>.Empty;
+
+        public override ImmutableArray<CustomModifier> RefCustomModifiers => ImmutableArray<CustomModifier>.Empty;
+
+        public sealed override ImmutableArray<TypeSymbol> TypeArguments => TypeParameters.Cast<TypeParameterSymbol, TypeSymbol>();
+
+        public sealed override int Arity => TypeParameters.Length;
 
-        public override ImmutableArray<CustomModifier> RefCustomModifiers
+        internal TypeParameterConstraintKind GetTypeParameterConstraints(int ordinal)
         {
-            get
-            {
-                return ImmutableArray<CustomModifier>.Empty;
-            }
+            var clause = this.GetTypeParameterConstraintClause(ordinal);
+            return (clause != null) ? clause.Constraints : TypeParameterConstraintKind.None;
         }
 
-        public sealed override ImmutableArray<TypeSymbol> TypeArguments
+        internal ImmutableArray<TypeSymbol> GetTypeParameterConstraintTypes(int ordinal)
         {
-            get
-            {
-                return TypeParameters.Cast<TypeParameterSymbol, TypeSymbol>();
-            }
+            var clause = this.GetTypeParameterConstraintClause(ordinal);
+            return (clause != null) ? clause.ConstraintTypes : ImmutableArray<TypeSymbol>.Empty;
         }
 
-        public sealed override int Arity
+        private TypeParameterConstraintClause GetTypeParameterConstraintClause(int ordinal)
         {
-            get
-            {
-                return TypeParameters.Length;
-            }
+            var clauses = TypeParameterConstraints;
+            return (clauses.Length > 0) ? clauses[ordinal] : null;
         }
 
-        internal sealed override bool TryGetThisParameter(out ParameterSymbol thisParameter)
+        protected virtual ImmutableArray<TypeParameterConstraintClause> TypeParameterConstraints => throw ExceptionUtilities.Unreachable;
```

Comment:
>protected virtual ImmutableArray<TypeParameterConstraintClause> TypeParameterConstraints => throw ExceptionUtilities.Unreachable; [](start = 8, length = 129)

It feels like this method should be abstract. It should be clear from this PR how each derived class implements it.